File: gpt.php
#!/usr/bin/php

<?php
// To install it globally, follow these steps:
// - sudo cp gpt.php /usr/local/bin/gpt
// - sudo chmod +x /usr/local/bin/gpt
// After that, you can run the script from anywhere by typing `gpt` in a new fresh terminal.

$cor_verde = "\033[0;32m";
$cor_amarela = "\033[0;33m";
$cor_vermelha = "\033[0;31m";
$cor_reset = "\033[0m";

// Variable name for the OpenAI API key
$envVarName = 'OPENAI_API_KEY';

// Check if the OpenAI API key is set as an environment variable
if (getenv($envVarName) === false) {
    echo $cor_vermelha . "The OpenAI API key is not set.\nYou can create an API key at https://platform.openai.com/api-keys.\nPlease enter the API key: ". $cor_reset . PHP_EOL;
    $handle = fopen("php://stdin", "r");
    $apiKey = trim(fgets($handle));

    if (!empty($apiKey)) {
        // Determine the shell configuration file based on the current shell
        $shellName = basename($_SERVER['SHELL']);
        $shellConfigFile = $_SERVER['HOME'] . '/.'.$shellName.'rc';

        switch (basename($_SERVER['SHELL'])) {
            case 'bash':
                $shellConfigFile = $_SERVER['HOME'] . '/.bashrc';
                break;
            case 'zsh':
                $shellConfigFile = $_SERVER['HOME'] . '/.zshrc';
                break;
            default:
                echo $cor_vermelha . "It was not possible to determine the current shell configuration file." . $cor_reset . PHP_EOL;
                exit;
        }

        // Add the API key to the shell configuration file
        file_put_contents($shellConfigFile, "\nexport $envVarName=\"$apiKey\"\n", FILE_APPEND);

        // Inform the user that the API key has been added and provide instructions to reload the configuration file
        echo $cor_vermelha . "The API Key was added to your $shellConfigFile. Please restart the terminal or run `source $shellConfigFile` to apply the changes." . $cor_reset . PHP_EOL;
        exit;
    } else {
        echo $cor_vermelha . "Invalid API Key. Can't perform the request." . $cor_reset . PHP_EOL;
        ;
        exit;
    }
}

$isCommit = false;
// Check if the script was called with an argument
if($argc > 1) {
    $prompt = $argv[1];
    // If the parameter is 'commit', the script will generate a commit message based on the git diff
    if($prompt == 'commit') {
        $isCommit = true;
        $diff = shell_exec('git diff');
        $prompt = "Write a concise semantic git commit message. Follow this format:
            1. The first line: A summary of the key changes, up to 60 characters, using conventional commits format.
            2. The second line: A blank line.
            3. The third line onwards: A detailed but concise description of the changes.

            Use the following git diff output as a reference for the commit message:\n: $diff";
    }
} else {
    echo $cor_verde . "Hi. How can I help you?". $cor_reset . PHP_EOL;
    $prompt = readline();
}


// Your OpenAI API key
$apiKey = getenv($envVarName);

// Request Header
$headers = [
    'Authorization: Bearer ' . $apiKey,
    'Content-Type: application/json'
];

// Data to be sent in the request
$data = [
    // 'model' => 'gpt-3.5-turbo',
    'model' => 'gpt-4-turbo',
    'messages' => [
        [
            'role' => 'system',
            'content' => 'You are an assistant that generates output for a Linux terminal. Please ensure all responses are formatted correctly and consider the limitations and capabilities of this environment when providing your response.',
        ],
        [
            'role' => 'user',
            'content' => $prompt,
        ]
    ],
    'temperature' => 0.7,
    'max_tokens' => 600  // Adjust the number of tokens as needed
];
if($isCommit) {
    $data['max_tokens'] = 400;
}
echo $cor_verde . "Wait..." . $cor_reset . PHP_EOL;

// API endpoint for OpenAI (you may need to modify this based on the specific API you are using)
$apiUrl = 'https://api.openai.com/v1/chat/completions';

// Initialize cURL session
$ch = curl_init($apiUrl);

// Set cURL options
curl_setopt($ch, CURLOPT_POST, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

// Execute the POST request
$response = curl_exec($ch);

// Close cURL session
curl_close($ch);

// Decode and display the response
$responseData = json_decode($response, true);
echo $cor_amarela . "=== Answer:" . $cor_reset . PHP_EOL;
if(isset($responseData['error'])) {
    echo "Erro: " . $responseData['error']['message'] . "\n";
    exit(1);
}

foreach ($responseData['choices'] as $choice) {
    $response = $choice['message']['content'];
    $response = str_replace('```', '', $response);

    echo $response . PHP_EOL;

    if($isCommit) {
        echo $cor_amarela . "Do you want to add, commit, and push with this message? (Y/n):" . $cor_reset;
        $confirmacao = trim(fgets(STDIN));  // Lê a entrada do usuário

        // Confirm the user's intention to add, commit, and push the response
        if (strtolower($confirmacao) === 'y' || empty($confirmacao)) {
            $escapedResponse = escapeshellarg($response);
            shell_exec("git add .");
            shell_exec("git commit -m $escapedResponse");
            shell_exec("git push");
        }
    }
    exit(0);
}

File: gpt.py
#!/usr/bin/env python3

import os
import sys
import subprocess
import json

# ANSI color codes
COR_VERDE = "\033[0;32m"
COR_AMARELA = "\033[0;33m"
COR_VERMELHA = "\033[0;31m"
COR_RESET = "\033[0m"

import os
import sys
import json
import subprocess
import requests

# Variable name for the OpenAI API key
env_var_name = 'OPENAI_API_KEY'

# Function to get the API key
def get_api_key():
    api_key = os.getenv(env_var_name)
    if api_key is None:
        print(f"{COR_VERMELHA}The OpenAI API key is not set.{COR_RESET}\nYou can create an API key at https://platform.openai.com/api-keys.\nPlease enter the API key: ")
        api_key = input().strip()
        if api_key:
            shell_name = os.path.basename(os.getenv('SHELL'))
            shell_config_file = os.path.join(os.getenv('HOME'), f'.{shell_name}rc')

            if shell_name == 'bash':
                shell_config_file = os.path.join(os.getenv('HOME'), '.bashrc')
            elif shell_name == 'zsh':
                shell_config_file = os.path.join(os.getenv('HOME'), '.zshrc')
            else:
                print(f"{COR_VERMELHA}It was not possible to determine the current shell configuration file.{COR_RESET}")
                sys.exit(1)

            with open(shell_config_file, 'a') as file:
                file.write(f'\nexport {env_var_name}="{api_key}"\n')

            print(f"{COR_VERDE}The API Key was added to your {shell_config_file}. {COR_AMARELA}Please restart the terminal or run `{COR_RESET}source {shell_config_file}{COR_AMARELA}` to apply the changes.{COR_RESET}")
            sys.exit(0)
        else:
            print(f"{COR_VERMELHA}Invalid API Key. Can't perform the request.{COR_RESET}")
            sys.exit(1)
    return api_key

# Function to interact with OpenAI API
def interact_with_openai(api_key, prompt, is_commit=False):
    headers = {
        'Authorization': f'Bearer {api_key}',
        'Content-Type': 'application/json'
    }
    data = {
        'model': 'gpt-4-turbo',
        'messages': [
            {
                'role': 'system',
                'content': 'You are an assistant that generates output for a Linux terminal. Please ensure all responses are formatted correctly and consider the limitations and capabilities of this environment when providing your response.'
            },
            {
                'role': 'user',
                'content': prompt
            }
        ],
        'temperature': 0.7,
        'max_tokens': 600 if not is_commit else 400
    }
    print(f"{COR_VERDE}Wait...{COR_RESET}")

    response = requests.post('https://api.openai.com/v1/chat/completions', headers=headers, data=json.dumps(data))
    response_data = response.json()

    if 'error' in response_data:
        print(f"{COR_VERMELHA}Error: {response_data['error']['message']}{COR_RESET}")
        sys.exit(1)

    for choice in response_data['choices']:
        response_text = choice['message']['content'].replace('```', '')
        print(f"{COR_AMARELA}=== Answer:{COR_RESET}")
        print(response_text)

        if is_commit:
            confirm = input(f"{COR_AMARELA}Do you want to add, commit, and push using this message? (Y/n): {COR_RESET}").strip().lower()
            if confirm == 'y' or confirm == '':
                subprocess.run("git add .", shell=True)
                commit_message = response_text.replace('\n', '\\n').replace('"', '\\"')
                subprocess.run(f'git commit -m "{commit_message}"', shell=True)
                subprocess.run("git push", shell=True)
        sys.exit(0)

# Main script execution
if __name__ == "__main__":
    api_key = get_api_key()
    is_commit = False

    if len(sys.argv) > 1:
        prompt = sys.argv[1]
        if prompt == 'commit':
            is_commit = True
            diff = subprocess.getoutput('git diff')
            if not diff:
                print(f"{COR_VERMELHA}No changes detected in git diff.{COR_RESET}\nPlease make changes before committing.")
                sys.exit(0)
            prompt = f"Write a concise semantic git commit message. Follow this format:\n1. The first line: A summary of the key changes, up to 60 characters, using conventional commits format.\n2. The second line: A blank line.\n3. The third line onwards: A detailed but concise description of the changes.\n\nUse the following git diff output as a reference for the commit message:\n: {diff}"
    else:
        print(f"{COR_VERDE}Hi. How can I help you?{COR_RESET}")
        prompt = input().strip()

    interact_with_openai(api_key, prompt, is_commit)

